### PR TITLE
Swift: move keypath dataflow writes to fix types

### PR DESCRIPTION
**Pull request by rdmarsh2**
This moves the read and write steps on keypath nodes so that the component node is the destination of the read and source of the write, rather than being the source of the read and destination of the write. This means that the type of the dataflow node matches the type that the read/write implies, which will be important when type pruning is added.

### CODE REVIEW

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll [hunk 1]
```diff
@@ -263,11 +263,18 @@ private module Cached {
     or
     // Flow from the unique parameter of a key path expression to
     // the first component in the chain.
-    nodeTo.(KeyPathComponentNodeImpl).getComponent() =
-      nodeFrom.(KeyPathParameterNode).getComponent(0)
+    exists(KeyPathExpr keyPath |
+      nodeFrom.(KeyPathComponentNodeImpl).getComponent() = keyPath.getComponent(keyPath.getNumberOfComponents() - 1) and
+      nodeTo.(KeyPathReturnNodeImpl).getKeyPathExpr() = keyPath
+    )
```
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:269]: Is the comment above this disjunct still correct now? It looks like this is now a step from the _last_ component to the return, right?

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll [hunk 2]
```diff
@@ -263,11 +263,18 @@ private module Cached {
...
     or
-    nodeFrom.(KeyPathComponentPostUpdateNode).getComponent() =
-      nodeTo.(KeyPathParameterPostUpdateNode).getComponent(0)
+    exists(KeyPathExpr keyPath |
+      nodeTo.(KeyPathComponentPostUpdateNode).getComponent() = keyPath.getComponent(keyPath.getNumberOfComponents()-1) and
+      nodeFrom.(KeyPathReturnPostUpdateNode).getKeyPathExpr() = keyPath
+    )
+    
+  //  nodeFrom.(KeyPathComponentPostUpdateNode).getComponent() =
+  //    nodeTo.(KeyPathParameterPostUpdateNode).getComponent(0)
```
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:277]: Should probably be deleted now

**Review by MathiasVP — COMMENTED** [review]
LGTM once the small comments are addressed (and we have a happy DCA run, of course)

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll (hunk 1 above)
Comment (rdmarsh2) [reply swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll]: Fixed

**Review by MathiasVP — APPROVED** [review]
LGTM!

**Pull request merged**